amundsen_application/static/js/ducks/bookmark/sagas.ts

```diff
@@ -0,0 +1,88 @@
+import { call, put, takeEvery } from 'redux-saga/effects';
+import { SagaIterator } from 'redux-saga';
+
+ import {
+  AddBookmark,
+  AddBookmarkRequest,
+  GetBookmarks,
+  GetBookmarksForUser,
+  GetBookmarksForUserRequest,
+  GetBookmarksRequest,
+  RemoveBookmark,
+  RemoveBookmarkRequest
+} from "./types";
+
+ import {
+  addBookmark,
+  removeBookmark,
+  getBookmarks,
+} from "./api/v0";
+
+
+ // AddBookmarks
+export function* addBookmarkWorker(action: AddBookmarkRequest): SagaIterator {
+  let response;
+  const { resourceKey, resourceType } = action;
+
+  try {
+    yield call(addBookmark, resourceKey, resourceType);
+
+    // TODO - Consider adding the newly bookmarked resource directly to local store. This would save a round trip.
+    response = yield call(getBookmarks);
+    yield put({ type: AddBookmark.SUCCESS, payload: response.bookmarks });
+  } catch(e) {
+    yield put({ type: AddBookmark.FAILURE, payload: response });
+   }
+}
+export function* addBookmarkWatcher(): SagaIterator {
+  yield takeEvery(AddBookmark.ACTION , addBookmarkWorker)
```

Comment:
On second thought with what I'm describing we would actually want `takeLeading`, not `takeLastest`.

Reply:
I'm not sure we should use `takeLeading` or `takeLatest` here. I can see the use case for clicking on the same bookmark, but if a user adds two different bookmarks this could lead to some weird edge cases where one is added and the other is stuck.

We don't have a good pattern established for this yet, but I think we could look into making better use of `isLoading` to disable buttons while an existing request is in flight. Maybe this could be a follow up task?

Reply:
Ah that's completely right. There is an approach for sharing actions between components involving appending unique prefixes or suffixes to the action type, in which case we could then use `takeLeading`, but that is its own endeavor. I don't think the edge case I raised is worth sinking time into that right now given our other work. I support resolving this comment and making use of `isLoading` later, or forgoing it and putting the effort into the Redux layer when we have the time -- your call!